core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.kt

```diff
@@ -0,0 +1,257 @@
+        EnvelopePhysics.getMechanicalEnergyConsumed(envelope, envelopePath, schedule.rollingStock)
+
+    return ResultTrain(
+        speeds, headPositions, stops, routeOccupancies, mechanicalEnergyConsumed, signalSightings, zoneUpdates
+    )
+}
+
+private fun routeOccupancies(
+    zoneOccupationChangeEvents: MutableList<ZoneOccupationChangeEvent>, rawInfra: SimInfraAdapter, envelope: Envelope
+): Map<String, ResultOccupancyTiming> {
+    val routeOccupancies = mutableMapOf<String, ResultOccupancyTiming>()
+    val zoneOccupationChangeEventsByRoute = mutableMapOf<String, MutableList<ZoneOccupationChangeEvent>>()
+    for (event in zoneOccupationChangeEvents) {
+        for (route in rawInfra.zoneMap.inverse()[event.zone]!!.routes) {
+            zoneOccupationChangeEventsByRoute.getOrPut(route.id) { mutableListOf() }.add(event)
+        }
+    }
+
+    for ((route, events) in zoneOccupationChangeEventsByRoute) {
+        events.sortBy { it.time }
+        val entry = events.first()
+        assert(entry.isEntry)
+        val exit = events.last()
+        if (exit.isEntry) {
+            routeOccupancies[route] = ResultOccupancyTiming(entry.time.toDouble() / 1000, envelope.totalTime)
+        }
+        routeOccupancies[route] = ResultOccupancyTiming(entry.time.toDouble() / 1000, exit.time.toDouble() / 1000)
+    }
+    return routeOccupancies
+}
+
+private data class ZoneOccupationChangeEvent(
+    val time: Long,
+    val offset: Distance,
+    val zoneIndexInPath: Int,
+    val isEntry: Boolean,
+    val blockIdx: Int,
+    val zone: ZoneId,
+)
+
+private fun zoneOccupationChangeEvents(
+    startOffset: Distance,
+    blockPath: StaticIdxList<Block>,
+    blockInfra: BlockInfra,
+    envelope: Envelope,
+    rawInfra: SimInfraAdapter,
+    trainLength: Double
+): MutableList<ZoneOccupationChangeEvent> {
+    var zoneCount = 0
+    var currentOffset = startOffset
+    val zoneOccupationChangeEvents = mutableListOf<ZoneOccupationChangeEvent>()
+    for ((blockIdx, block) in blockPath.withIndex()) {
+        for (zonePath in blockInfra.getBlockPath(block)) {
+            // Compute occupation change event
+            if (currentOffset > envelope.endPos.meters) break
+            val entryOffset = maxOf(0.meters, currentOffset)
+            val entryTime = envelope.interpolateTotalTimeMS(entryOffset.meters)
+            val zone = rawInfra.getNextZone(rawInfra.getZonePathEntry(zonePath))!!
+            zoneOccupationChangeEvents.add(
+                ZoneOccupationChangeEvent(
+                    entryTime, entryOffset, zoneCount, true, blockIdx, zone
+                )
+            )
+            currentOffset += rawInfra.getZonePathLength(zonePath)
+            if (currentOffset > envelope.endPos.meters) {
+                zoneCount++
+                break
+            }
+            val exitOffset = maxOf(0.meters, currentOffset + trainLength.meters)
+            if (exitOffset <= envelope.endPos.meters) {
```

Comment:
It seems to be handled when computing the projection, every signal goes back to VL after the last event.

Generally we assume that trains "disappear" at the end of their final stop duration, to avoid "ghost trains" messing with the signals. Realistically the path should include the part where it moves out of the station into storage, but that doesn't sound convenient for the users. If we consider that the trains stay there, then the signals shouldn't go back to VL at all. I don't think it should be related to the projection.

We could discuss it in teams, it doesn't seem that straightforward.

Reply:
https://github.com/DGEXSolutions/osrd/issues/3731